enhancements/network/bgp-ovn-kubernetes.md

```diff
@@ -0,0 +1,366 @@
+reply traffic. Traditionally this is done by the gateway having its own operator to detect pod subnets via kubernetes
+API. With BGP, this is no longer necessary. A user can simply configure RouteAdvertisements, and the pod subnet routes
+will be dynamically learned by an external gateway capable of BGP peering.
+
+##### Egress IP
+
+EgressIP feature that is dynamically moved between nodes. By enabling the feature in RouteAdvertisements, OVN-Kubernetes
+will automatically change frr-k8s configuration so that the node where an egress IP resides will advertise the IP. This
+eliminates the need for nodes to be on the same layer 2 segment, as we no longer have to rely on gratuitous ARP (GARP).
+
+##### Services
+
+LoadBalancer and External IP are supported by MetalLB. This feature will handle advertising cluster IPs for services.
+Note, the usefulness of advertising ClusterIP may be limited as it will be advertised by multiple nodes (similar to
+anycast) and may not work well with stateful connections.
+
+##### Egress Service
+
+Full support.
+
+##### Egress Firewall
+
+Full support.
+
+##### Egress QoS
+
+Full Support.
+
+
+##### Network Policy/ANP
+
+Full Support.
+
+##### Direct Pod Ingress
+
+Direct pod ingress is already enabled for the default cluster network. With direct pod ingress, any external entity can
+talk to a pod IP if it sends the packet to the node where the pod lives. Previously, support for direct pod ingress on
+user-defined networks was not supported. RouteAdvertisements with this enhancement may select user-defined networks and
+enable pod network advertisements. Therefore, it only makes sense to also accept direct pod ingress for these
+user-defined networks as well, especially since we know the selected subnets will not overlap in IP address space.
+
+Modifications will need to be made so that for packets arriving on br-ex, flows are added to steer the selected pod
+subnets to the right OVN GR patch port. Additionally, modifications will need to be made in the kernel routing table to
+leak routes to the pod subnets from each user-defined VRF into the default VRF routing table.
+
+#### Deployment Considerations
+
+##### MetalLB
+
+MetalLB is supported today as a day 2 add-on operator. MetalLB supports FRR mode, and has recently added support for
+FRR-K8S mode. When MetalLB is deployed in FRR-K8S mode, it handles instantiating FRR-K8S. For the purposes of
+OVN-Kubernetes integration, we need this as a day 0 function. For pods relying on the underlay and not using an overlay
+encapsulation, the BGP routes will need to be learned by the time the pods come up.
+
+In order to achieve day 0 functionality, FRR-K8S will be launched by CNO as a host networked pod on each node (where
+applicable). MetalLB will be modified so that it can launch in FRR-K8S mode, without deploying FRR-K8S, and integrate
+with the one launched by CNO.
```

Comment:
```suggestion
with the one launched by CNO.

This means, there will be two ways of operating:

- The current behaviour, which is: CNO does not deploy FRR-K8s, the MetalLB Operator deploys both MetalLB and FRR-K8s
- CNO deploys FRR-K8s and MetalLB leverages the FRR-K8s instance deployed by CNO

## MetalLB deploys FRR-K8s

This is the behaviour implemented in 4.16 as Tech Preview and planned to be GA in 4.17. The users don't care about pod to pod reacheability with BGP, they want to use MetalLB and possibly FRR-K8s for learning routes on the nodes.
The MetalLB operator will deploy FRR-K8s and MetalLB in the same namespace.

## CNO deploys FRR-K8s

In this version, CNO deploys FRR-K8s in a separate namespace so it can be used for the extra features described in this proposal.

The user will need to set:

- a flag on the CNO configuration, to deploy FRR-K8s and to instruct OVN-K of the presence of the FRR controller
- a deployment mode "FRR-K8s already present" in the "MetalLB" resource, to instruct the operator to deploy MetalLB in FRR-K8s mode but without deploying FRR-K8s

In this scenario the permissions given to MetalLB over the FRR-K8s resources must be changed from namespace scoped to cluster scoped.

## Handling configuration errors

Because of this dual way of deploying, we must be careful in handling the scenario where a user mistakenly deployed FRR-K8s both from the MetalLB Operator and from CNO.

In this scenario, the FRR-K8s instance of CNO must take precedence and the MetalLB Operator must produce an error as soon as it sees the extra FRR-K8s instance, and possibly un-deploy the MetalLB pods so it's clear some exceptional scenario is happening.
```

Reply:
Getting back on this, as we (I) missed one detail, which is, the CRDs are deployed upfront as part of the bundle. This means, when deploying from CNO and trying to deploy metallb, OLM will cry as they are already there. Need to think about it as I don't have a solution upfront.

Reply:
@trozet can you unresolve?